platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/type/AbstractVariantSelectionByAttributesFailure.java

```diff
@@ -16,27 +16,29 @@
 
 package org.gradle.internal.component.resolution.failure.type;
 
+import org.gradle.api.artifacts.capability.CapabilitySelector;
 import org.gradle.api.artifacts.component.ComponentIdentifier;
 import org.gradle.api.internal.attributes.AttributeContainerInternal;
 import org.gradle.api.internal.attributes.ImmutableAttributes;
 import org.gradle.api.internal.catalog.problems.ResolutionFailureProblemId;
-import org.gradle.internal.component.external.model.ImmutableCapabilities;
 import org.gradle.internal.component.resolution.failure.interfaces.VariantSelectionByAttributesFailure;
 
+import java.util.Set;
+
 /**
  * An abstract {@link VariantSelectionByAttributesFailure} that represents the situation when a variant
  * was requested via variant-aware matching and that matching failed.
  */
 public abstract class AbstractVariantSelectionByAttributesFailure extends AbstractResolutionFailure implements VariantSelectionByAttributesFailure {
     private final ComponentIdentifier targetComponent;
     private final ImmutableAttributes requestedAttributes;
-    private final ImmutableCapabilities requestedCapabilities;
+    private final Set<CapabilitySelector> capabilitySelectors;
```

Comment:
❗ This should be an immutable set to preserve the immutability of the failure heirarchy.